sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * A case-insensitive string-to-string map, which is used to represent data source options.
+ */
+public class DataSourceV2Options {
+  private Map<String, String> keyLowerCasedMap = new HashMap<>();
+
+  /**
+   * Adds one more entry to the options.
+   * This should only be called by Spark, not data source implementations.
+   */
+  public void addOption(String key, String value) {
```

Comment:
The check added for addOption protects modifying the options passed to the datasource, but data source can still add new options by accident. I think it might be safer to pass DataSourceV2Options that are Unmodifiable by the data source.

Reply:
good point, I'll make it immutable.